lib/algos/cache/GapSet.ts

```diff
@@ -0,0 +1,352 @@
+        if (gap && gap.lastKey < firstKey) {
+            // closest gap found is not overlapping, ignore it
+            gap = null;
+        }
+        if (gap) {
+            if (gap.lastKey >= lastKey) {
+                // return fully overlapping gap already cached
+                return { gap, weightToMerge: 0 };
+            }
+            if (gap.lastKey === firstKey) {
+                // new gap is chained with last gap: merge the full weight value
+                weightToMerge = weight;
+            } else if (gap.firstKey === firstKey) {
+                // last gap is fully contained in the new gap: remove the
+                // last gap's weight from the weight to merge
+                weightToMerge = Math.max(weight - gap.weight, 0);
+            } else {
+                // last gap is not fully contained nor chained: default to returning
+                // the new weight (heuristic)
+                weightToMerge = weight;
+            }
+            if (weight > this._maxWeight) {
+                // if the new gap's weight exceeds the max, split the existing gap found in
+                // two chained gaps, by using the same key as a link (this is important
+                // to ensure that lookupGap() can return them as a single contiguous gap)
+                gap = {
+                    firstKey: gap.lastKey,
+                    // split gap is temporarily single-key, will be extended by setGap()
+                    lastKey: gap.lastKey,
+                    // split gap only contains the weight not already included in the previous gap
+                    //weight: Math.max(weight - gap.weight, 0),
+                    weight: 0,
+                };
+                // there may be an existing gap starting with 'lastKey': delete it first
+                this._gaps.delete(gap);
+                // then add the new split gap to be extended thereafter
+                this._gaps.add(gap);
+                return { gap, weightToMerge };
+            }
+        } else {
+            // create a new single-key gap that will be extended to 'lastKey' in setGap()
+            // during the merge process
+            gap = {
+                firstKey,
+                lastKey: firstKey,
+                weight: 0,
+            };
+            this._gaps.add(gap);
+            weightToMerge = weight;
+        }
+        return { gap, weightToMerge };
+    }
+
+    /**
+     * Record a gap between two keys, associated with a weight to limit
+     * individual gap sizes in the cache.
+     *
+     * The function handles splitting and merging existing gaps to
+     * maintain an optimal weight of cache entries.
+     *
+     * @param {string} firstKey - first key of the gap
+     * @param {string} lastKey - last key of the gap, must be greater
+     * or equal than 'firstKey'
+     * @param {number} weight - total weight between 'firstKey' and 'lastKey'
```

Comment:
I haven't looked at the implementation of the caller code but I'm guessing that it reduces all of the keys (the DMs and versions) into the `weight` somewhere before calling. What do you think of passing the entire list of keys instead, that way we can actually compute the exact weight of each of the gaps that are created even if the caller is unaware of how we are actually storing them.